src/modules/rlm_imap/rlm_imap.c

```diff
@@ -0,0 +1,301 @@
+
+extern fr_dict_autoload_t rlm_imap_dict[];
+fr_dict_autoload_t rlm_imap_dict[] = {
+	{ .out = &dict_radius, .proto = "radius" },
+	{ .out = &dict_freeradius, .proto = "freeradius" },
+	{ NULL }
+};
+
+
+
+extern fr_dict_attr_autoload_t rlm_imap_dict_attr[];
+fr_dict_attr_autoload_t rlm_imap_dict_attr[] = {
+	{ .out = &attr_auth_type, .name = "Auth-Type", .type = FR_TYPE_UINT32, .dict = &dict_freeradius },
+	{ .out = &attr_user_name, .name = "User-Name", .type = FR_TYPE_STRING, .dict = &dict_radius },
+	{ .out = &attr_user_password, .name = "User-Password", .type = FR_TYPE_STRING, .dict = &dict_radius },
+	{ .out = &attr_reply_message, .name = "Reply-Message", .type = FR_TYPE_STRING, .dict = &dict_radius },
+	{ NULL }
+};
+
+static int rlm_imap_cmp(UNUSED void *instance, REQUEST *request, UNUSED VALUE_PAIR *thing, VALUE_PAIR *check,
+			UNUSED VALUE_PAIR *check_pairs, UNUSED VALUE_PAIR **reply_pairs)
+{
+	rad_assert(check->vp_type == FR_TYPE_STRING);
+
+	RINFO("imap-Paircmp called with \"%pV\"", &check->data);
+
+	if (strcmp(check->vp_strvalue, "yes") == 0) return 0;
+	return 1;
+}
+
+/*
+ *	Do any per-module initialization that is separate to each
+ *	configured instance of the module.  e.g. set up connections
+ *	to external databases, read configuration files, set up
+ *	dictionary entries, etc.
+ */
+static int mod_instantiate(void *instance, UNUSED CONF_SECTION *conf)
+{
+	rlm_imap_t	*inst = instance;
+	if (paircmp_register_by_name("imap-Paircmp", attr_user_name, false, rlm_imap_cmp, inst) < 0) {
+		return -1;
+	}
+	return 0;
+}
+
+/*
+ *	Find the named user in this modules database.  Create the set
+ *	of attribute-value pairs to check and reply with for this user
+ *	from the database. The authentication code only needs to check
+ *	the password, the rest is done here.
+ */
+static rlm_rcode_t CC_HINT(nonnull) mod_authorize(void *instance, UNUSED void *thread, REQUEST *request)
+{
+	VALUE_PAIR 	*vp;
+	rlm_imap_t	*inst		= instance;
+	
+	/*
+	 *Check to see that we have a username and password
+	 *print out the username
+	 *if a value is missing return NOOP
+	 */
+	vp = fr_pair_find_by_da(request->packet->vps, attr_user_name, TAG_ANY);
+	if (vp == NULL) return RLM_MODULE_NOOP;
+	RDEBUG2("Authorizing: %pP", vp);
+	vp = fr_pair_find_by_da(request->packet->vps, attr_user_password, TAG_ANY);
+	if (vp == NULL) return RLM_MODULE_NOOP;
+	/*
+	 *Check to see if there is already a section that will be working on the request
+	 */
+	if (!module_section_type_set(request, attr_auth_type, inst->auth_type)) return RLM_MODULE_NOOP;
+	
+	return RLM_MODULE_OK;
+}
+
+
+
+
+
+//TODO: Add server url
+//TODO: Setup SSL verification options
+
+/*
+ *Called when the IMAP server responds
+ *It checks if the response was CURLE_OK
+ *If it wasn't we returns REJECT, if it was we returns OK
+*/
+static rlm_rcode_t mod_authenticate_resume(UNUSED void *instance, UNUSED void *thread, UNUSED REQUEST *request, void *rctx)
+{
+	fr_curl_io_request_t     	*randle    = rctx;
```

Comment:
Use talloc_get_type_abort when assigning from void.  See other examples in the server.